domain/src/main/scala/org/parsec/protocol/Parsec.scala

```diff
@@ -0,0 +1,125 @@
+      |                                 | | | |
+      |                       ETHEREUM     or    BITCOIN or ..
+    """.stripMargin
+
+  val DEFAULT_CHANNEL = "default"
+  val ALLOWED_CURRENCIES = Set("ETH", "BTC")
+
+  /**
+    * The seller needs to produce an INVOICE (through NFC or Parsec channel)
+    *
+    * @param invoiceAddress the sellect address for the currency to be transferred to
+    * @param price          the amount of currency to exchange
+    * @param currency       ETH or BTC supported
+    * @param invoiceType    (optional) use to pass-in other metadata about the invoice
+    */
+  case class Invoice(invoiceAddress: String, price: Double, currency: String, invoiceType: String = "") {
+    assert(ALLOWED_CURRENCIES.contains(currency), s"PasecProtocol: Currency $currency is not an allowed currency. Use: " + ALLOWED_CURRENCIES.mkString(" or "))
+
+    // The Hash can be calculated as a transformation function to the invoiceAddress
+    // TODO: implement me
+    def produceInvoiceHash = "######-####-####"
+  }
+
+  /**
+    * The gist of the Parsec Protocol lies here!
+    *
+    * We need to ensure that the <b>order</b> of Signed Invoices is guaranteed
+    * No cluster of distributed systems can guarantee that (due to varying network latencies etc)
+    *
+    * So within the Parsec network - we need to preserve a chain of HashPointers that provide ordering guarantees
+    */
+  case class HashPointer(transactionID: String, // i.e. invoiceID
+                         transactionHash: String)
+
+  /**
+    * The consumer needs to settle/pay the INVOICE
+    *
+    * @note step-2a
+    */
+  case class SignedInvoice(invoiceID: String = "auto-generate-random-UUID+channel", // import java.util.UUID
+                           invoiceSignature: String,
+                           supplierAddress: String, // no need to know PK
+                           buyerPublicKey: String,
+                           buyerAddress: String,
+                           currency: String,
+                           price: Double,
+                           channel: String = DEFAULT_CHANNEL,
+                           hashPointerToPrevious: HashPointer) {
+    assert(ALLOWED_CURRENCIES.contains(currency), s"PasecProtocol: Currency $currency is not an allowed currency. Use: " + ALLOWED_CURRENCIES.mkString(" or "))
+    assert(buyerPublicKey.length equals 128, "Wrong public key format. An ECDSA public key is 128 hex characters long")
+    assert(isValidBuyerAddress)
+
+    def isValidBuyerAddress: Boolean = {
+      if (currency == "ETH")
+        return buyerAddress equals ethereumAddressFromPublicKey(buyerPublicKey)
+      else
+        return ((buyerAddress equals bitcoinP2PKHAddressFromPublicKey(buyerPublicKey))
+               ||
+               (buyerAddress equals bitcoinP2SHAddressFromPublicKey(buyerPublicKey)))
+    }
+  }
+
+  def ethereumAddressFromPublicKey(key: String): String = {
```

Comment:
I think we should introduce one or more `traits` EthereumEncryption BitcoinEncryption etc - to allow mix-in of capabilities